feat: Token Network Filter UI [Mobile]

gambinish opened the pull request:
## **Description**

This PR introduces token network filter UI component. It lives behind a feature flag `PORTFOLIO_VIEW` in order to allow this to get merged while backend changes are in flight.

Integration with the multichain asset list will occur in a separate PR. We'll likely want to add additional e2e tests when that happens.

### Running this branch

`yarn && yarn setup`
`PORTFOLIO_VIEW=1 yarn watch`
`yarn start:ios` or `i` from the watch process

Please ensure that this PR looks okay both with and without the feature flag running. I have introduced a horizontal scroll view because there simply was not enough screen real estate to make the network filter look okay with truncated text.

## **Related issues**

Fixes:

## **Manual testing steps**

1. Run app with feature flag
2. Go to main Portfolio view
3. Pressing the network filter should trigger the bottom sheet, allowing the user to select different filter options.

## **Screenshots/Recordings**

https://github.com/user-attachments/assets/636d1627-5da3-4826-83a1-924ff4f28603

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

gambinish committed:
feat: Add sortAssets utility function

gambinish committed:
chore: Breakup monolithic Tokens/index.tsx into smaller components

gambinish committed:
fix: Correctly handle null check for buyable footer button

gambinish committed:
fix: Tweak boolean logic for rendering footer components

gambinish committed:
chore: Update snapshot

gambinish committed:
chore: Type out tokens as TokenI[]

gambinish committed:
chore: Remove unecessary comments around import statements in Networth component

gambinish committed:
Merge branch 'main' into chore/componetize-tokens-screen

gambinish committed:
Merge branch 'main' into chore/componetize-tokens-screen

gambinish committed:
Merge branch 'main' into chore/componetize-tokens-screen

gambinish committed:
chore: Add unit tests to utility file, provide exports from index file

gambinish committed:
chore: Rename handleBalance to me more descriptive of what its logic is handling

gambinish committed:
chore: rename Networth to PortfolioBalance for clarity

gambinish committed:
chore: Cleanup styles extra whitespace

gambinish committed:
Merge branch 'main' into chore/componetize-tokens-screen

gambinish committed:
fix: Update snapshot tests

gambinish committed:
Merge branch 'chore/componetize-tokens-screen' of github.com:MetaMask/metamask-mobile into chore/componetize-tokens-screen

gambinish committed:
fix: Remove unused component

gambinish committed:
Merge branch 'main' into chore/componetize-tokens-screen

gambinish committed:
chore: Move PotfolioBalance above TabView

gambinish committed:
fix: Wrap PortfolioBalance and ScrollableView in fragment for token rendering

gambinish committed:
fix: Padding tweak below tab bar for consistency of token list

gambinish committed:
fix: Reorganize test suite for PortfolioBalance

gambinish committed:
fix: Revert changes to wallet test

gambinish committed:
fix: Update fiatBalance text style to DisplayMD

gambinish committed:
chore: Patch PreferencesController with tokenSortConfig

gambinish committed:
chore: Patch preferences controller with type

gambinish committed:
feat: Build token sorting UI, patch PreferencesController

gambinish committed:
chore: Merge main, address merge conflicts

gambinish committed:
fix: ListHeaderComponent not needed in FlatList anymore

gambinish committed:
fix: Update snapshots and lint

gambinish committed:
chore: Merge develop, address conflicts

gambinish committed:
chore: Remove destructive index

gambinish committed:
Merge branch 'main' into feat/mmassets_357-sort-import-tokens-mobile--reorg-portfolio-balance

gambinish committed:
chore: Merge main, address conflicts

gambinish committed:
chore: Repatch preferences controller with tokenSortConfig

gambinish committed:
fix: Declning balance sort

gambinish committed:
fix: Export raw value from deriveBalanceFromAssetMarketDetails

gambinish committed:
fix: deriveBalanceFromAssetMarketDetails.test.ts

gambinish committed:
fix: Update unit tests and snapshots

gambinish committed:
fix: Lint unused imports from Tokens/index.tsx

gambinish committed:
chore: Update testIds and locale strings

gambinish committed:
fix: Update WalletView import selector

gambinish committed:
fix: Update snapshots

gambinish committed:
chore: Cleanup

gambinish committed:
fix: Add navigation type

salimtb committed:
fix: fix snapshot

salimtb committed:
Merge branch 'main' into feat/mmassets_357-sort-import-tokens-mobile--reorg-portfolio-balance

gambinish committed:
chore: Update snapshots

gambinish committed:
Merge branch 'feat/mmassets_357-sort-import-tokens-mobile--reorg-portfolio-balance' into feat/mmassets_357-sort-import-tokens-mobile--token-sort-config

gambinish committed:
fix: Update placeholders in switch statement

gambinish committed:
Merge branch 'feat/mmassets_357-sort-import-tokens-mobile--token-sort-config' of github.com:MetaMask/metamask-mobile into feat/mmassets_357-sort-import-tokens-mobile--token-sort-config

gambinish committed:
chore: Add additional e2e test for import token footer link

gambinish committed:
chore: Update snapshots

gambinish committed:
chore: Patch PreferencesController with tokenNetworkFilter

gambinish committed:
fix: Update to correct type in PreferenceController patch

gambinish committed:
fix: Correct method name in patched PreferenceController

gambinish committed:
chore: Integrate Filter dropdown with PreferencesController

gambinish committed:
fix: Break out of switch statement rather than returning arbitrary string

gambinish committed:
chore: Init BottomSheet refactor

gambinish committed:
fix: Remove arbitrary 8 second delays from import-tokens.spec

gambinish committed:
fix: Use ButtonBase component to customize button text color more directly

gambinish committed:
fix: Update snapshot

gambinish committed:
fix: Mock initial state that PortfolioBalance component needs

gambinish committed:
fix: Mock tabs in snapshot

gambinish committed:
fix: Consider Error boundary during Wallet rendering process

gambinish committed:
fix: Update localized string

gambinish committed:
Merge branch 'feat/mmassets_357-sort-import-tokens-mobile--reorg-portfolio-balance' into feat/mmassets_357-sort-import-tokens-mobile--token-sort-config

gambinish commented:
WIP recording:

https://github.com/user-attachments/assets/66a59148-4ff8-4b45-bf97-39d4fee05f0f

gambinish committed:
Merge branch 'main' into feat/mmassets_357-sort-import-tokens-mobile--reorg-portfolio-balance

gambinish committed:
Merge branch 'feat/mmassets_357-sort-import-tokens-mobile--reorg-portfolio-balance' into feat/mmassets_357-sort-import-tokens-mobile--token-sort-config

gambinish committed:
Merge branch 'main' into feat/mmassets_357-sort-import-tokens-mobile--reorg-portfolio-balance

gambinish committed:
Merge branch 'feat/mmassets_357-sort-import-tokens-mobile--reorg-portfolio-balance' into feat/mmassets_357-sort-import-tokens-mobile--token-sort-config

gambinish committed:
Merge branch 'main' into feat/mmassets_357-sort-import-tokens-mobile--token-sort-config

gambinish committed:
chore: Add TokenSortBottomSheet to RootModalFlow in Routes

gambinish committed:
fix: Resolve upstream conflicts

gambinish committed:
fix: Lint:tsc

gambinish committed:
fix: Lint

gambinish committed:
chore: Init unit tests for new UI

gambinish committed:
chore: Merge main

gambinish committed:
chore: Show selected network in filter

gambinish committed:
fix: Update unit tests and snapshots

gambinish committed:
chore: Merge main, address conflicts

gambinish committed:
fix: Add test for refresh control on tokens screen

gambinish committed:
chore: Update testID selectors for token list

gambinish committed:
chore: Add tokenSort BottomSheet trigger test

gambinish committed:
chore: Lint

gambinish committed:
fix: Update snapshots

gambinish committed:
chore: TokenSortBottomSheet unit tests

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
chore: Cleanup

gambinish committed:
chore: Cleanup, and queryByTestID for test stability

gambinish committed:
chore: More cleanup

gambinish committed:
chore: Merge main, address conflicts

gambinish committed:
fix: Lint

gambinish committed:
Merge branch 'feat/mmassets_357-sort-import-tokens-mobile--token-sort-config--bottom-sheet' into feat/mmassets-431_network-filter-ui

gambinish committed:
fix: Lint after conflicts

gambinish committed:
fix: Missing testIDs after merge for unit tests

gambinish committed:
chore: Merge main, address conflicts

gambinish committed:
refactor: Filter ActionSheet should be BottomSheet

gambinish committed:
chore: Tweak styles to fix merge conflict

gambinish committed:
fix: Update unit tests with correct import path

gambinish committed:
chore: Add unit tests for TokenFilterBottomSheet

gambinish committed:
chore: Add locale strings

gambinish committed:
chore: Add MM_TOKEN_NETWORK_FILTER_ENABLED feature flag

gambinish committed:
fix: Add missing locale string

gambinish committed:
chore: Update snapshots

gambinish committed:
fix: Cleanup localisation strings and code comments

gambinish committed:
fix: Cleanup localisation strings, remove code comments, update feature flag to PORTFOLIO_VIEW

gambinish commented:
Realizing now, that this is not using the `ListItemSelect` as it should. I'll make the update shortly, and it should make this PR smaller.

gambinish committed:
chore: Revert changes to component-library that are no longer needed

gambinish committed:
fix: Lint whitespace diff

gambinish committed:
fix: Another whitespace diff fix

gambinish commented on patches/@metamask+preferences-controller+11.0.0.patch:
```diff
@@ -19,11 +19,12 @@ index 30e985c..ec9db71 100644
 +  showMultiRpcModal: { persist: false, anonymous: false },
-+  tokenSortConfig: { persist: true, anonymous: false }
++  tokenSortConfig: { persist: true, anonymous: false },
++  tokenNetworkFilter: { persist: true, anonymous: false },
```
@salimtb here is the PR to core: https://github.com/MetaMask/core/pull/4875

I added you as a reviewer there as well 👍

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
chore: Merge main

gambinish committed:
fix: Remove outdated patch to preferenceController

gambinish committed:
chore: Repatch package with tokenNetworkFilter

gambinish committed:
fix: Lint

gambinish committed:
chore: Merge main, address conflicts

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
chore: Merge main

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish commented:
@metamaskbot update-policies

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
refactor: Update AssetListControls buttons to icons per design revisions

gambinish committed:
fix: Update button text wrapping with snapshots

gambinish committed:
Merge branch 'feat/mmassets-431_network-filter-ui' of github.com:MetaMask/metamask-mobile into feat/mmassets-431_network-filter-ui

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
fix: Lint

gambinish committed:
Merge branch 'feat/mmassets-431_network-filter-ui' of github.com:MetaMask/metamask-mobile into feat/mmassets-431_network-filter-ui

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish commented:
Just leaving a comment per @Cal-L suggestion. The failing Bitrise workflow is related to some flaky Android tests that are currently being investigated. They are unrelated to this PR and it can be reviewed.

gambinish committed:
chore: Replace unrelated build flag

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

Cal-L commented on app/components/UI/Tokens/index.tsx:
```diff
@@ -218,25 +232,61 @@ const Tokens: React.FC<TokensI> = ({ tokens }) => {
   const onActionSheetPress = (index: number) =>
     index === 0 ? removeToken() : null;
 
+  const isTokenFilterEnabled = process.env.PORTFOLIO_VIEW === '1';
```
```suggestion
  const isTokenFilterEnabled = process.env.PORTFOLIO_VIEW === 'true';
```
Let's keep the flag values consistent

Cal-L commented on app/components/UI/Tokens/styles.ts:
```diff
@@ -158,11 +158,37 @@ const createStyles = (colors: Colors) =>
+    },
+    controlButtonText: {
+      color: colors.text.default,
+      ...fontStyles.normal,
```
Use font from theme instead

gambinish replied:
It turns out I didn't need this fontStyle after all. I removed it here: https://github.com/MetaMask/metamask-mobile/pull/11808/commits/ab20b3eb1ca2702419d5409b5a4d9571ebb9a784

Cal-L commented on patches/@metamask+preferences-controller+13.1.0.patch:
```diff
@@ -11,18 +11,19 @@ index 97182f2..634de2e 100644
 +    useSafeChainsListValidation: { persist: true, anonymous: true },
 +    showMultiRpcModal: { persist: false, anonymous: false },
 +    tokenSortConfig: { persist: true, anonymous: false },
++    tokenNetworkFilter: { persist: true, anonymous: false },
```
Has this logic been updated in the controllers already?

gambinish replied:
Good call, it may be updated now, I'll check. I think only the `tokenNetworkFilter` has not yet been released in core

gambinish replied:
Okay, they are added in 13.2 I can bump that package here as well.

gambinish replied:
Okay, bumped and repatched here: https://github.com/MetaMask/metamask-mobile/pull/11808/commits/33d6d30b759fc1881a1f641075a64963811e148f

gambinish replied:
It looks like some of these were included in the latest release of PreferencesController. Are you okay with addressing this upgrade in a separate PR? I don't want to expand the scope of this any more than it needs to be.

darkwing replied:
Bump on Nick's question @Cal-L

Cal-L replied:
Sounds good, let's make sure there's a ticket to track it

Cal-L reviewed (commented):
Left some comments. Let's also remove the empty horizontal gutter from the pills.

gambinish committed:
fix: Remove unused font style

gambinish committed:
fix: Use true for feature flag

gambinish committed:
fix: Unblock podfile during CI

gambinish committed:
chore: Merge main, fix conflicts

vinnyhoward committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

Cal-L reviewed (commented):
@gambinish I couldn't tell, did we also remove the padding white space on the sides of the scrollable pills? For example, the cut off to the right of this pic

<img width="225" alt="image" src="https://github.com/user-attachments/assets/12b9bc62-b600-412b-b8e7-46fa7459e241">

gambinish committed:
fix: Tweak padding

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

gambinish committed:
chore: Update snapshot

gambinish committed:
Merge branch 'feat/mmassets-431_network-filter-ui' of github.com:MetaMask/metamask-mobile into feat/mmassets-431_network-filter-ui

sethkfman committed:
feat: 7.34.0 (#11858)

This is the release candidate for version 7.34.0
 
  # Team sign-off checklist
  - [ ] team-accounts
  - [ ] team-assets
  - [x] team-confirmations
  - [x] team-design-system
  - [ ] team-lavamoat
  - [x] team-mobile-platform 
  - [x] team-notifications
  - [x] team-ramp
  - [ ] team-sdk
  - [ ] team-security
  - [x] team-snaps-platform 
  - [x] team-stake
  - [ ] team-tiger
  - [ ] team-wallet-framework
  - [ ] team-wallet-ux


  # Reference
- Testing plan sheet -
https://docs.google.com/spreadsheets/d/1tsoodlAlyvEUpkkcNcbZ4PM9HuC9cEM80RZeoVv5OCQ/edit?gid=404070372#gid=404070372

---------

Signed-off-by: Kai Huang <kai.huang@consensys.net>
Co-authored-by: runway-github[bot] <73448015+runway-github[bot]@users.noreply.github.com>
Co-authored-by: MetaMask Bot <37885440+metamaskbot@users.noreply.github.com>
Co-authored-by: metamaskbot <metamaskbot@users.noreply.github.com>
Co-authored-by: tommasini <46944231+tommasini@users.noreply.github.com>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Xiaoming Wang <7315988+dawnseeker8@users.noreply.github.com>
Co-authored-by: Nico MASSART <NicolasMassart@users.noreply.github.com>
Co-authored-by: Xiaoming Wang <dawnseeker8@gmail.com>
Co-authored-by: Nicholas Ellul <nicholas.ellul1@gmail.com>
Co-authored-by: legobt <6wbvkn0j@anonaddy.me>
Co-authored-by: OGPoyraz <omergoktugpoyraz@gmail.com>
Co-authored-by: Salim TOUBAL <salim.toubal@outlook.com>
Co-authored-by: Pedro Pablo Aste Kompen <wachunei@gmail.com>
Co-authored-by: legobeat <109787230+legobeat@users.noreply.github.com>
Co-authored-by: tommasini <tommasini15@gmail.com>
Co-authored-by: Bryan Fullam <bryanfullam@gmail.com>
Co-authored-by: Mpendulo Ndlovu <mpendulo@elefantel.com>
Co-authored-by: Kaihuang72490 <147628638+Kaihuang72490@users.noreply.github.com>
Co-authored-by: Cal-L <cleun007@gmail.com>
Co-authored-by: Cal Leung <cal.leung@consensys.net>
Co-authored-by: sahar-fehri <sahar.fehri@consensys.net>
Co-authored-by: Vince Howard <vincenguyenhoward@gmail.com>
Co-authored-by: Matthew Walsh <matthew.walsh@consensys.net>

gambinish committed:
Merge branch 'main' into feat/mmassets-431_network-filter-ui

Cal-L reviewed (approved):
LGTM
